core/src/main/scala/quasar/qscript/MapFunc.scala

```
@@ -164,20 +164,66 @@ object MapFunc {
         })
   }
 
+  // we start from the leaves, and make the inner parts into Constant as we go
+  // so the immediate child is always a constant, else we know the endpoint was not a constant
+  // use freeTransCata!
+  def foldConstant[T[_[_]]: Recursive: Corecursive, A]:
+      CoMFR[T, A] => Option[CoMFR[T, A]] =
+    _.run.fold(
+      κ(None),
+      {
+        // relations
+        case And(
+            Embed(CoEnv(\/-(Constant(Embed(ejson.Common(ejson.Bool(v1))))))),
+            Embed(CoEnv(\/-(Constant(Embed(ejson.Common(ejson.Bool(v2)))))))) =>
+          CoEnv[A, MapFunc[T, ?], T[CoEnv[A, MapFunc[T, ?], ?]]](
+            Constant[T, T[CoEnv[A, MapFunc[T, ?], ?]]](
+              CommonEJson.inj(ejson.Bool[T[EJson]](v1 && v2)).embed).right).some
```

It’s not important for this PR, but constant folding should be able to fail … or at least return `Undefined`. So, in this case, something like
```scala
case And(
  Embed(CoEnv(\/-(Constant(Embed(const1))))),
  Embed(CoEnv(\/-(Constant(Embed(const2)))))) =>
  (ejson.Common.prj(const1) ⊛ ejson.Common.prj(const2))(…).getOrElse(CoEnv(Undefined().right))
```
We don’t want to fail _immediately_, because it’s possible that something like `IfUndefined(x, y)` catches the statically-known `Undefined()` and gets rewritten to `y`.